flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/pipeline/EstimatorBase.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.flink.ml.pipeline;
+
+import org.apache.flink.ml.api.core.Estimator;
+import org.apache.flink.ml.api.misc.param.Params;
+import org.apache.flink.ml.common.MLEnvironmentFactory;
+import org.apache.flink.ml.operator.batch.BatchOperator;
+import org.apache.flink.ml.operator.batch.source.TableSourceBatchOp;
+import org.apache.flink.ml.operator.stream.StreamOperator;
+import org.apache.flink.ml.operator.stream.source.TableSourceStreamOp;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.TableEnvironment;
+import org.apache.flink.table.api.internal.TableImpl;
+import org.apache.flink.table.api.java.StreamTableEnvironment;
+
+/**
+ * The base class for estimator implementations.
+ *
+ * @param <E> The class type of the {@link EstimatorBase} implementation itself
+ * @param <M> class type of the {@link ModelBase} this Estimator produces.
+ */
+public abstract class EstimatorBase<E extends EstimatorBase<E, M>, M extends ModelBase<M>>
+	extends PipelineStageBase<E> implements Estimator<E, M> {
+
+	public EstimatorBase() {
+		super();
+	}
+
+	public EstimatorBase(Params params) {
+		super(params);
+	}
+
+	@Override
+	public M fit(TableEnvironment tEnv, Table input) {
+		MLEnvironmentFactory.get(getMLEnvironmentId()).setTableEnvironment(tEnv, input);
+		return fit(input);
+	}
+
+	/**
+	 * Train and produce a {@link ModelBase} which fits the records in the given {@link Table}.
+	 *
+	 * @param input the table with records to train the Model.
+	 * @return a model trained to fit on the given Table.
+	 */
+	public M fit(Table input) {
```

Comment:
Does this method have to be public?

Reply:
Thanks, this method is more convenient with the MLEnvironmentFactory, so I think it should be public